app/shared/components/Wallet/Panel/Form/Stake.js

```diff
@@ -29,91 +30,220 @@ export default class WalletPanelFormStake extends Component<Props> {
+    const parsedNetWeight = net_weight.split(' ')[0];
 
     this.state = {
-      cpuAmount: new Decimal(parsed_cpu_weight),
-      cpuOriginal: new Decimal(parsed_cpu_weight),
-      netAmount: new Decimal(parsed_net_weight),
-      netOriginal: new Decimal(parsed_net_weight)
+      decimalCpuAmount: Decimal(parsedCpuWeight),
+      cpuOriginal: Decimal(parsedCpuWeight),
+      decimalNetAmount: Decimal(parsedNetWeight),
+      netOriginal: Decimal(parsedNetWeight),
+      confirming: false,
+      formError: null,
+      submitDisabled: true
     };
   }
 
+  onSubmit = (e) => {
+    if (!this.state.submitDisabled) {
+      this.setState({
+        confirming: true
+      });
+    }
+    e.preventDefault();
+    return false;
+  }
+
+  onKeyPress = (e) => {
+    if (e.key === 'Enter') {
+      this.onSubmit(e);
+
+      e.preventDefault();
+      return false;
+    }
+  }
+
+  onError = (error) => {
+    let errorMessage;
+
+    if (error !== true) {
+      errorMessage = error;
+    }
+
+    this.setState({
+      submitDisabled: true,
+      formError: errorMessage
+    });
+  }
+
+  onChange = (name, value) => {
+    const decimalFieldName = `decimal${name.charAt(0).toUpperCase()}${name.slice(1)}`;
+
+    this.setState({
+      submitDisabled: false,
+      formError: null,
+      [decimalFieldName]: Decimal(value)
+    });
+  }
+
+  onBack = () => {
+    this.setState({
+      confirming: false
+    });
+  }
+
+  onConfirm = () => {
+    const {
+      account,
+      actions
+    } = this.props;
+
+    const {
+      decimalCpuAmount,
+      decimalNetAmount
+    } = this.state;
+
+    const {
+      setStake
+    } = actions;
+
+    this.setState({
+      confirming: false
+    });
+
+    setStake(account, decimalNetAmount, decimalCpuAmount);
+  }
+
   render() {
     const {
       account,
-      actions,
       balance,
       onClose,
-      system,
-      validate
+      validate,
+      t
     } = this.props;
 
     const {
+      decimalCpuAmount,
       cpuOriginal,
-      netOriginal
+      decimalNetAmount,
+      netOriginal,
+      submitDisabled
     } = this.state;
 
     const EOSbalance = balance.EOS || 0;
 
-    const lastTransactions = [];
-    if (
-      system.DELEGATEBW_LAST_TRANSACTION
-      && system.DELEGATEBW_LAST_TRANSACTION.transaction_id
-    ) {
-      lastTransactions.push(system.DELEGATEBW_LAST_TRANSACTION);
-    }
-    if (
-      system.UNDELEGATEBW_LAST_TRANSACTION
-      && system.UNDELEGATEBW_LAST_TRANSACTION.transaction_id
-    ) {
-      lastTransactions.push(system.UNDELEGATEBW_LAST_TRANSACTION);
-    }
+    const shouldShowConfirm = this.state.confirming;
+    const shouldShowSuccess = validate.STAKE === 'SUCCESS' && !shouldShowConfirm;
+    const shouldShowError = validate.STAKE === 'FAILURE' && !shouldShowConfirm;
+    const shouldShowForm = validate.STAKE === 'NULL' && !shouldShowSuccess && !shouldShowError && !shouldShowConfirm;
+    const errorObject = validate.STAKE_LAST_ERROR;
```

Comment:
All of these validate calls should be using `system.*` after the actions are changed.

Reply:
ok so should I add a system/stake action and handle everything in there?

Reply:
Yeah I think so. It already handles almost all of these cases already (success, pending, error, returns the error).